### PR TITLE
Force JSON request for getting paste data & password retry

### PR DESCRIPTION
**Pull request by rugk**
This PR fixes https://github.com/PrivateBin/PrivateBin/issues/174 and fixes https://github.com/PrivateBin/PrivateBin/issues/159 and fixes https://github.com/PrivateBin/PrivateBin/issues/215
This disallows/prevents the transmission of the paste data (which is actually more than just the "ciphertext" and was wrongly labelled) in the HTML file and instead always uses a JSON request to get the paste data before decrypting it.

## Changes

* The password prompt logic has been improved and now should not show an error message in the background
* As for https://github.com/PrivateBin/PrivateBin/issues/159 it thereby always sends one JSON request no matter whether encryption succeeded/failed or whether the burnafterreading option is enabled. That said, one could of course still analyse the size of the JSON request and see that the paste is e.g. expired if the request is quite small.
* When a paste with burnafterreading enabled is received from the server (via JSON) it is instantly deleted.
* Thus MITM people trying to do things will not see the ciphertext, or – if they do it intently then they trigger a paste deletion and users will notice it
* No UI changes yet, for users everything stays the same
* I did *not* made it configurable as it is a relevant for security and there are not much alternative client implementations for PrivateBin yet, we could break (it is, of course, however a breaking change)
* Most bots should be blocked, because they do not execute JavaScript
* If there are really bots, which parse JS we can blacklist them in the JS source code
* To fix https://github.com/PrivateBin/PrivateBin/issues/215 this also adds an retry button

## Reasoning
Of course, we cannot completely protect any message interception if the PrivateBin links is sent over insecure channels (an attacker could create a new Bin and replace the link), but that does not mean we should not try to make it as hard as possible for an attacker if this is realistic and does not introduce a bad UI or similar disavantages…

This change mostly protects against passive attacks, or "attackers", who did not want to weaken the system intentionally, but just used curl or so to download pages and e.g. store them in some kind of cache or website preview, etc. (think of IRC, Skype,  or other bots…)

## ToDo

* [ ] adjust unit tests

### CODE REVIEW

**Comment by rugk** [comment]
@elrido Please merge these changes with https://github.com/PrivateBin/PrivateBin/commit/f54036976a486d3f7c83e52fd37ea357fd4e7a62. I'd argue for removing the option as explained above – there are no clients who we could have to stay backwards-compatible to…
Also in my PR here I did the deletion at a different position, which makes sense combined with the fact that I removed the support for getting the paste data in a HTML.

File: lib/PrivateBin.php [hunk 1]
```diff
@@ -373,6 +376,11 @@ private function _read($dataid)
                     unset($data->meta->salt);
                 }
                 $this->_data = json_encode($data);
+
```
Comment (elrido) [inline lib/PrivateBin.php]: Please move this logic into the Paste object. The controller should not contain model logic.

File: js/privatebin.js [hunk 2]
```diff
@@ -43,10 +43,21 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
...
+        var BadBotUA = [
+            'Bot',
+            'bot'
+        ];
+
+        /**
          * character to HTML entity lookup table
          *
          * @see    {@link https://github.com/janl/mustache.js/blob/master/mustache.js#L60}
-         * @name Helper.entityMap
```
Comment (elrido) [inline js/privatebin.js]: The @name are required for the JSDoc to work with our JS object pattern.

File: js/privatebin.js [hunk 3]
```diff
@@ -43,10 +43,21 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
         var me = {};
 
         /**
+         * list of UserAgents (parts) known to belong to a bot
+         *
+         * @private
+         * @enum   {Object}
+         * @readonly
+         */
+        var BadBotUA = [
```
Comment (elrido) [inline js/privatebin.js]: Which bots do execute JS and are targeted by these pattern? (Asking so we can replicate this in the unit tests)

File: js/privatebin.js [hunk 4]
```diff
@@ -721,32 +758,53 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
...
+                if (typeof callback === 'function') {
+                    return callback(pasteData);
+                }
+
+                // alternatively just using inline
+                return pasteData;
+            }
+
+            // reload data
+            Uploader.prepare();
```
Comment (elrido) [inline js/privatebin.js]: It seems a bit odd to use the Uploader object to download data from the API. Should that object maybe be renamed?

File: js/privatebin.js [hunk 5]
```diff
@@ -1333,8 +1390,8 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
             if (pasteMetaData.burnafterreading) {
                 // display paste "for your eyes only" if it is deleted
 
-                // actually remove paste, before we claim it is deleted
-                Controller.removePaste(Model.getPasteId(), 'burnafterreading');
+                // the paste has been deleted when the JSOn with the ciohertext
```
Comment (elrido) [inline js/privatebin.js]: Just some typos: JSO**N**, ci**p**hertext

File: js/privatebin.js (hunk 2 above)
Comment (rugk) [inline js/privatebin.js]: But these are private, internal variables so they actually do not have any name. You cannot use them from the "outside". The only valid name would probably be "entityMap", but we do not have to document it via "@name"…

File: js/privatebin.js (hunk 3 above)
Comment (rugk) [inline js/privatebin.js]: I have not tested it actually. The strings here were merely templates, which could target many bots…
Maybe better remove it and wait for issues being opened, when one discovers such a bot…
Then we can add it to this blacklist.

File: js/privatebin.js (hunk 4 above)
Comment (rugk) [inline js/privatebin.js]: It's done in other parts of the code too. Works well, but yes, could be renamed. It's just a wrapper around jQueries ajax requests. (Strictly speaking every ajax request is an upload, of course, but well…)
However, I think we can do this once this is merged, as it does not only affect this code part.

**Comment by rugk** [comment]
As for the unit tests I would have to get them to work locally anyway, see https://github.com/PrivateBin/PrivateBin/issues/218. That's why I've never looked into them.
Basically everything, which is broken is intentionally. As far as I see only the PHP tests are broken and I think you can easily fix them, i.e. the change is only minimal and quite similar to https://github.com/PrivateBin/PrivateBin/commit/f54036976a486d3f7c83e52fd37ea357fd4e7a62, where you've already added a unit test, which should test all this stuff…

File: js/privatebin.js (hunk 4 above)
Comment (elrido) [inline js/privatebin.js]: Ok, let's do this change in a separate commit before the next release.

File: js/privatebin.js (hunk 2 above)
Comment (elrido) [inline js/privatebin.js]: The default JSDoc run as documented in the README will not document private methods and variables. But adding the option `-p` or `--access all` will include these private symbols. Its not ultra urgent but ideally I would like to have the option to have these documented, too.

**Comment by elrido** [comment]
Will work on this next during the coming week.

**Comment by elrido** [comment]
I merged master in and updated the unit tests for both PHP and JS. There is one regression that I found and tried to solve but failed so far: The `testReadExpired()` test PrivateBinTest.php fails, because we now don't actually read the paste anymore in the HTML view. This means that neither this error works which would have come from an Exception thrown on getting the paste, nor is the automatic expiration triggered until the subsequent JSON API request from the JS.

@rugk: Could you please double check the changes I did in the merge and ensure that the expiration error is getting displayed by the JS logic (a link to the line in privatebin.js would help me)?

File: lib/PrivateBin.php [hunk 6]
```diff
@@ -369,12 +366,15 @@ private function _read($dataid)
         try {
             $paste = $this->_model->getPaste($dataid);
             if ($paste->exists()) {
-                $data              = $paste->get();
-                $this->_doesExpire = property_exists($data, 'meta') && property_exists($data->meta, 'expire_date');
-                if (property_exists($data->meta, 'salt')) {
-                    unset($data->meta->salt);
+                // reading paste is only possible via JSON call
+                if ($this->_request->isJsonApiCall()) {
```
Comment (rugk) [inline lib/PrivateBin.php]: Hmm, if this check fails, it does not result in an error, does it? (I know, my version before also did not, but mhh…)

Maybe to clarify the code we could also do early returns here,  i.e. don't nest if statements, but return them. But this is only a minor idea, and seeing that there is some error logic below, it does not seem possible so easy.

File: lib/PrivateBin.php [hunk 7]
```diff
@@ -147,10 +147,7 @@ public function __construct()
                 );
                 break;
             case 'read':
-                // reading paste is disallowed in HTML display
```
Comment (rugk) [inline lib/PrivateBin.php]: I think this should stay here. As we use JSON requests for reading the paste, it needs to be only done when reading…
https://github.com/PrivateBin/PrivateBin/pull/216/commits/183ebe518b02c52b22d2ad5f472dbfcb862c83f2#diff-e0c90c7248782346a7f382fd588d1039R150

But ah, you mofved that logic into the `_read` function, okay.

File: tpl/bootstrap.php [hunk 8]
```diff
@@ -524,18 +524,18 @@
 				</div>
 			</footer>
 		</main>
-		<div id="serverdata" class="hidden" aria-hidden="true">
 <?php
 if ($DISCUSSION):
 ?>
+		<div id="serverdata" class="hidden" aria-hidden="true">
```
Comment (rugk) [inline tpl/bootstrap.php]: Oh, yeah good idea. Basically it seems the "serverdata" is actually no data anymore, but just the comment templates.

So we could remove the whole if thing and get that part static, too. But taht's a thing to do later…

**Comment by rugk** [comment]
Okay, so I had a look at the source and it seems to be okay. I also tested the expiration with 5 minutes and it works (even counts down the seconds) and I tested the expiration with auto-expire.

The expiration is still handled in PHP, this PR should not change much about this. The `$this->isBurnafterreading` check is done in the `get` function, so seems to work. (When I opened this PR it was in the `_read` function, but that was changed, as it seems.

**Comment by rugk** [comment]
I am on a regression…

**Comment by rugk** [comment]
Uups, my last commit was in the wrong baranch, sry

**Comment by elrido** [comment]
Ah, wait a second: I just corrected `testReadExpired()` to use the API, but I just realized that the `testReadNonexisting()` still checks the HTML output and doesn't fail. That is an information leak, since non-existing and expired pastes need to behave identical.

**Comment by elrido** [comment]
Had to remove a feature regarding the hiding of the clone button when on an expiring paste. It wasn't working anymore, since it relied on reading the paste during for the template rendering. This was discovered due to a warning of scrutinizer about the unused _data variable.
